engine/class_modules/paladin/sc_paladin_holy.cpp

```diff
@@ -273,13 +281,37 @@ struct holy_shock_heal_t : public paladin_heal_t
   }
 };
 
+// Glimmer of light stuff
+
+struct glimmer_of_light_damage_t : public paladin_spell_t
+{
+  glimmer_of_light_damage_t( paladin_t* p ) : paladin_spell_t( "glimmer_of_light", p, p->find_spell( 325984 ) )
+  {
+    background = true;
+  }
+};
+
+struct glimmer_of_light_heal_t : public paladin_spell_t
+{
+  glimmer_of_light_heal_t( paladin_t* p ) : paladin_spell_t( "glimmer_of_light", p, p->find_spell( 325983 ) )
+  {
+    background = true;
+  }
+};
+
+
 struct holy_shock_t : public paladin_spell_t
 {
   holy_shock_damage_t* damage;
   holy_shock_heal_t* heal;
   bool dmg;
 
-  holy_shock_t( paladin_t* p, util::string_view options_str )
+  glimmer_of_light_damage_t* glimmer_damage;
+  glimmer_of_light_heal_t* glimmer_heal;
+
+  std::vector<player_t*> glimmer_targets;
```

Comment:
Storing glimmer targets in holy_shock_t means you'll have one target list per Holy Shock line.
My suggestion here would be to:
1) Store the array/vector/queue/whatever you want it to be on the paladin_t level
2) Transfer the handle_glimmer() functions to paladin_t as trigger_glimmer() and apply_glimmer() (clearer names imho than handle_glimmer() and handle_glimmer_targets()
3) Inside apply_glimmer, make sure that: a) on glimmer application: add target to the vector and trigger the buff on that target. If the vector is full, remove the oldest target from the vector and expire its buff. b) on glimmer refresh, make sure the position of the target in the vector is updated to reflect the new application
4) Inside the glimmer target data buff's stack_change_callback, remove the target from the vector if the debuff expires naturally.

nb: I don't think stack_change_callback lets you differentiate between manually expired buffs (from apply_glimmer) and normally expired buff, so you may have to make it a specific glimmer_buff_t with an expire_override() override as this function receives the remaining duration (and stack count) on the buff at time of expiration.